src/basic/copy.c

```diff
@@ -386,14 +386,13 @@ static int fd_copy_directory(
                         continue;
                 }
 
-                if (buf.st_dev != original_device)
-                        continue;
-
                 if (S_ISREG(buf.st_mode))
                         q = fd_copy_regular(dirfd(d), de->d_name, &buf, fdt, de->d_name, override_uid, override_gid, copy_flags);
-                else if (S_ISDIR(buf.st_mode))
+                else if (S_ISDIR(buf.st_mode)) {
+                        if (buf.st_dev != original_device)
+                                continue;
```

Comment:
Right. I made my change to be minimal while keeping the current behaviour. But when you come over such a mount point, it's true that the underlying filesystem has at least an empty directory for the mount to be possible. So arguably the behaviour of `cp --one-file-system` is more logical. Shall I update my patch to behave like cp ?